### ISSUE TITLE
CMS field overrides with _schema.json

### ISSUE THREAD

**Issue by jimafisk** [plentico/plenti#204]
The discoverable CMS looks at the content source to provide a reasonable editing experience for users out-of-the-box without doing any special setup. However, there are going to be scenarios when a developer may want to override those defaults to provide a specific editing experience. They may also want to provide additional features that are difficult to discover from simple JSON:
- required fields
- select lists, multiselect lists, radio buttons, checkboxes
- autocomplete / entity reference
- text, number, boolean, date
- WYSIWYG (configurable options?)
- assets (image, doc)
- defined components (that aren't part of default content)

We should add an optional `_schema.json` file that lives inside Type folders (e.g. `content/pages`) that allows devs to do these overrides. This will require defining a specific API that must be followed. We could look to https://json-schema.org/ for inspiration.

**Comment by jimafisk** [comment]
Components (arrays) are probably the most complicated field type we need to deal with. Here's an example of how that could look:
 
<details>
<summary>Defaults</summary>

```json
{
  "components": [
    {
      "name": "hero",
      "fields": {
        "title": "This is a Hero",
        "image": {
          "src": "/assets/myimage.jpg",
          "alt": "Alternative text for image"
        }
      }
    },
    {
      "name": "grid",
      "fields": {
        "title": "This is a Grid",
        "items": [
          "item 1",
          "item 2",
          "item 3"
        ]
      }
    }
  ]
}
```
</details>

<details>
<summary>Schema</summary>

```json
{
  "components": {
    "_type": "components",
    "_value": [
      {
        "name": "hero",
        "fields": {
          "title": "This is a Hero",
          "image": {
            "src": "/assets/myimage.jpg",
            "alt": "Alternative text for image"
          }
        }
      },
      {
        "name": "grid",
        "fields": {
          "title": "This is a Grid",
          "items": [
            "slide 1",
            "slide 2",
            "slide 3"
          ]
        }
      }
    ],
    "_options": [
      {
        "name": "hero",
        "fields": {
          "title": "This is a Hero",
          "image": {
            "src": "/assets/myimage.jpg",
            "alt": "Alternative text for image"
          }
        }
      },
      {
        "name": "grid",
        "fields": {
          "title": "This is a Grid",
          "items": [
            "item 1",
            "item 2",
            "item 3"
          ]
        }
      },
      {
        "name": "slider",
        "fields": {
          "title": "This is a Slider",
          "items": [
            "item 1",
            "item 2",
            "item 3"
          ]
        }
      }
    ]
  }
}
```

</details>

The schema above is setting the default values to display both a `hero` and `grid` component, but there will be a widget to add new components of type `hero`, `grid`, or `slider`.

You could even define schema for nested components. So for instance if we look at the `grid` component, we could add tags to it:

<details>
<summary>Defaults</summary>

```json
{
  "name": "grid",
  "fields": {
    "title": "This is a Grid",
    "items": [
      {
        "title": "item 1",
        "tags": "tag A"
      },
      {
        "title": "item 2",
        "tags": "tag A, Tag B"
      },
      {
        "title": "item 3",
        "tags": "tag C"
      }
    ]
  }
}
```
</details>

<details>
<summary>Schema</summary>

```json
{
  "name": "grid",
  "fields": {
    "title": "This is a Grid",
    "items": {
      "_type": "components",
      "_value": [
        {
          "title": "item 1",
          "tags": "tag A"
        },
        {
          "title": "item 2",
          "tags": "tag A, Tag B"
        },
        {
          "title": "item 3",
          "tags": "tag C"
        }
      ],
      "_options": [
        {
          "title": "New item",
          "tags": {
            "_type": "checkboxes",
            "_value": "tag A",
            "_options": [
              "tag A",
              "tag B",
              "tag C"
            ]
          }
        }
      ]
    }
  }
}
```
</details>

**Comment by jimafisk** [comment]
In classic fashion I've gone back to the original plan on this. Currently I'm working on letting you define a `_schema.json` file separately from you `_blueprint.json` (which would be better if renamed `_defaults.json`). It seemed like putting these all in the same file was more confusing than just separating the concerns cleanly. The rough API so far looks like this:

<details>
<summary>Checkboxes</summary>

Saves value as array

Options = Listed checkboxes (set initial values as array in `_blueprints.json` (soon to be `_defaults.json`))

_schema.json:

```json
{
  "myfield": {
    "type": "checkbox", 
    "options": [
      "pizza",
      "chips",
      "soda",
      "cookie"
    ]
  }  
}   
```

</details>

<details>
<summary>Radio buttons</summary>

Saves as a string

Options = Listed radio buttons (set initial value as string in `_blueprints.json` (soon to be `_defaults.json`)

Example _schema.json:

```json
{
  "myfield": {
    "type": "radio", 
    "options": [
      "yes",
      "no",
      "maybe"
    ]
  }  
}   

```

</details>

<details>
<summary>WYSIWYG Editor</summary>

Saves as a string

Options = Rich text buttons ("all" can be used to add every button without listing them out)

```json
{
  "myfield": {
    "type": "wysiwyg", 
    "options": [
      "bold",
      "italic",
      "bullets",
      "clear"   
    ]
  }  
}   
```

</details>

Things to note:
- There are no values in `_schema.json`, if you want to set defaults use `_blueprint.json` (soon to be `_defaults.json`)
- Still haven't figured out components. This is complicated for a variety of reasons and I'd like you to be able to define them with minimal configuration
- If you want to target nested values, you use dot notation in your schema key (e.g. `"target.nested.key"`). The actual JSON in `_schema.json` is flat (just the first level, no nesting) so that's the only way to target nested keys. Anything that's nested beyond the first level is the actual widget configuration.
- All widgets should have an optional `required` key that you can set to `true`
- Still need to allow case insensitive like "WYSIWYG" vs "wysiwyg"

Ideally the API will stay very similar between the widgets for consistency, but as this gets fleshed out more I'm sure there will be cases where that would be difficult or misleading. I'll try to post updates as this evolves.

**Comment by jimafisk** [comment]
We could consolidate `checkbox` and `radio` into a single type called [choice](https://english.stackexchange.com/questions/439224/hypernym-for-radio-button-and-checkbox). Then have a `multiple` key that can be set to `true` to convert the default radio widget into a checkbox. We similarly could have a single type called `select` with a `multiple` key that allows a multi-select list. Alternatively we could have an optional `limit` key that defaults to no limit (checkboxes) down to a limit of `1` (radio buttons). It would allow more optionality for allowing a checkbox where you only want to allow selecting 3 of 5 available options at a time for example. Although you could omit this key entirely to remove the limit, we'd probably also want a value that would represent "no limit" as well (usually this is `-1`). Limit could be used other widgets like components I imagine.

**Comment by jimafisk** [comment]
Just spitballing for components... 

<details>
<summary>_schema.json</summary>

```json
"components": {
  "type": "components",
  "limit": -1,
  "options": [
    "hero",
    "grid",
    "slider",
    "cta"
  ]
}
```
</details>

Inside your content type folder (e.g. `content/pages`) you would have a special named folder: `_components/`. The options from the `_schema.json` above reference component folders nested in this folder:
- `content/pages/_components/hero/`
- `content/pages/_components/grid/`
- `content/pages/_components/slider/`
- `content/pages/_components/cta/`

Each component would have its own `_defaults.json` and `_schema.json`:
- `content/pages/_components/hero/_defaults.json`
- `content/pages/_components/hero/_schema.json`

Considerations:
- Nested components start to make my head spin, so let's ignore those for now. Potentially could keep nesting `_components` folders within each other.
- What if we want to share components across content types? Should components exist at the top-level `content/` folder? Maybe do this optionally to make them global, but they can also be put in a content type folder to scope them specifically to that type.
- It would be duplicating the default json for the initial components in the top-level `_defaults.json` file

**Comment by jimafisk** [comment]
Autocomplete / (Entity) Reference widget: Allows you to look up content that exists on the site and create a relationship to it.

<details>
<summary>Example _schema.json</summary>

```json
{
  "myfield": {
    "type": "autocomplete",
    "options": [
      {
        "type": "pages",
        "search": [
          "title",
          "image"
        ],
        "result": "filename"
      },
      {
        "type": "blog",
        "search": [
          "author"          
        ],
        "result": "filename"
      }
    ]  
  }
} 
```

</details>

Keys:
- `type` = Content type
- `search` = fields you want to search on
- `result` = what gets save to the field

**Comment by jimafisk** [comment]
Here's video tutorial explaining how to use `_schema.json` in your projects: https://youtu.be/wje7VzE7BXE

**Comment by jimafisk** [comment]
The field types from the video ^ should be ready for testing in [v0.5.13](https://github.com/plentico/plenti/releases/tag/v0.5.13). That does _not_ include components and keep in mind the API for these might change still.

**Comment by jimafisk** [comment]
Just wanted to mention the other field types discussed in https://github.com/plentico/plenti/issues/213:

<details>
<summary>ID</summary>

```json
"my_id_field": {
    "type": "id"
},
```

</details>

<details>
<summary>Untracked</summary>

```json
"my_private_field": {
    "type": "text",
    "before": "my_other_field"
},
```

</details>

**Comment by jimafisk** [comment]
More fields will continue to be added overtime, and someday maybe you'll even be able to define your own (https://github.com/plentico/plenti/issues/219). 

However, for now the schema types mentioned previously on this issue should be available so I'm going to close this issue.

Component schemas (https://github.com/plentico/plenti/issues/224) were implemented in [v0.5.16](https://github.com/plentico/plenti/releases/tag/v0.5.16) as well.
